File: c++/FindKthLargest.cpp
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        int left = 0, right = nums.size() - 1;
        while(true){
            int pos = partition(nums, left, right);
            if (pos == k-1) return nums[pos];
            if (pos > k-1) right = pos - 1;
            else left = pos + 1;
        }
    }
    
    int partition(vector<int>& nums, int left, int right){
        int pivot = nums[left];
        int l = left+1, r = right;
        while(l<=r){
            // this need to be first condition other wise we may need use continue to avoid stay in same while 
            if (nums[l]<pivot && nums[r]>pivot){
                swap(nums[l], nums[r]);
                l++;
                r--;
            }
            if (nums[l] >= pivot) l++;
            if (nums[r] <= pivot) r--;
                       
        }
        swap(nums[left], nums[r]);
        return r;
    }
};

class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        priority_queue<int, vector<int>, greater<int>> Q;
        int n = nums.size();
        for (int i = 0; i < n; i++) {
            if (i < k) {
                Q.push(nums[i]);
            }
            else {
                if(Q.top() <= nums[i]) {
                    Q.push(nums[i]);
                    Q.pop();
                }
            }
        }
        return Q.top();
    }
};

File: python/findKthLargest.py
class Solution:
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        left = []
        right = []
        middle = []
        pivot = nums[0]
        for num in nums:
            if num > pivot:
                left.append(num)
            elif num < pivot:
                right.append(num)

            else:
                middle.append(num)
        r = len(right)
        m = len(middle)
        l = len(left)
        if l >= k:
            return self.findKthLargest(left, k)

        elif l + m >= k:
            return pivot
        else:
            return self.findKthLargest(right, k - m - l)


class Solution:
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        queue = []
        for num in nums:
            heapq.heappush(queue, num)
            if len(queue) > k:
                heapq.heappop(queue)

        return queue[0]  # or heapq.heappop(queue)
